### DISCUSSION TITLE
Declarative Configuration in a lua file

### DISCUSSION THREAD

**Discussion by dorongold** [Kong/kong#7283] [Ideas and feature requests]
### Summary

Currently Kong supports writing configuration in a file, in YAML or JSON.
To provide a more flexible configuration, especially for those running Kong in different environments (dev, staging, production), an option to write configuration in `lua` can be very useful.

### Additional Details

There are many use-cases that come to mind:

- Using environment variables inside the declarative configuration, something like:
```
return {
services = {
  name = my-upstream
  url = os.getenv("UPSTREAM_URL_FOR_THIS_ENVRIONMENT")
   ...
  }
}
```
- Injecting parts of the configuration from a file, a Kubernetes config-map, or any other source which has different content in different environments.
This can be done using lua functions such as `dofile`, `load`, `loadfile`, `loadstring`.

- etc.

**Comment by dorongold** [comment]
@hishamhm currently (after the issue you referenced is merged) this piece of code in `kong/db/declarative/init.lua` still blocks the use of Lua declarative configuration:
```
 -- do not accept Lua by default
  accept = accept or { yaml = true, json = true }
```
In addition, when you guys decide to fix that and support Lua declarative configuration, it would be nice to change this line `setfenv(chunk, {})` and instead of setting an empty environment, at least pass in `_G.os`, so that `os.getenv` can be used inside the declarative configuration.

It would be nice the be updated here if and when this support is going to be added.

**Comment by mohatagarvit** [comment]
Should I make a different PR for these changes or continue in the last PR?
```
 -- do not accept Lua by default
  accept = accept or { yaml = true, json = true, lua = true }
```

`setfenv(chunk, {_G.os})` in the elseif of lua

**Comment by andrepiske** [comment]
@dorongold I'm currently trying to put together a tool that allows to configure Kong in a declarative way, pretty much like you described here.

It's not open sourced yet, but I plan to do it in the foreseeable future.

I have a question though: Is your idea only for Kong configurations in the DB-less mode?

**Comment by dorongold** [comment]
@andrepiske We only use DB-less mode, so that is the only use-case I was aiming at.
Not sure what you mean about putting together a tool.
Declarative configuration using a Lua file is almost there.
All it takes is that someone from the Kong team merges the tiny 2 line change in the attached PR (or for them to create a similar PR themselves and merge it).

**Comment by andrepiske** [comment]
@dorongold Oh that might be it, you're using DB-less mode.
What I mean by the tool is that I'm creating a tool to declare configuration (services, routes, plugins _etc_) of Kong in a declarative way, similar to [decK](https://deck.yolo42.com/). But I think it wouldn't be appropriate for db-less mode :)
